### PR TITLE
Feature: Operations histogram tooltip with data buckets

### PR DESCRIPTION
**Pull request by jhegg**
# problem statement

The Operations Histogram (formerly BarChart) needs the tooltip to represent the counts of normal/warning/critical. It might be nice to allow the user of the library to stuff that into the tooltip, but this histogram feels rather specific, so we are opting to hardcode the tooltip behavior. This will fix issue #79.

# solution

This is a breaking change. We're renaming the classes to reflect their specific use case, and altering the inputs to match our desired data model. We will then change the tooltip contents.

# discussion

@ggascoigne and I are pairing on this change. It's not yet done, but we've gotten through most of the breaking change refactoring.

### CODE REVIEW

**Comment by wa11-e** [comment]
Woohoo! new surge deployment available for viewing! :tada: [surgereview24775195.surge.sh](http://surgereview24775195.surge.sh)

**Comment by wa11-e** [comment]
Woohoo! new surge deployment available for viewing! :tada: [surgereview24780539.surge.sh](http://surgereview24780539.surge.sh)

**Comment by jhegg** [comment]
I think this PR is close to being finished for now. We need to decide if 1000+ operations is a thing, and if so, increase the tooltip width and adjust all the text/icon positions when that happens.

I'm not really sure what to do in the short term about the tooltip positioning versus allowing the histogram to have different height/width values, which makes it look sloppy. At the current sizing in the style guide the tooltip fits well, but if you increase the height/width you'll see what I'm talking about as it drifts away from what was previously an anchor point.

Current ideal histogram size used in style guide:
image

Original bar chart style guide size with new tooltip positioning, less than ideal:
image


We could:
- Leave it as-is until someone needs it to behave differently.
- Expose all the positioning attributes, which would mean several additional props. We could keep the current values as defaults, though. Then someone could fix the position if it's not behaving on their end.
- Do a better job of using relative/scaled units compared to the size of the histogram.

File: src/components/charts/OperationsHistoryBarChart/OperationsHistoryBarChart.jsx [hunk 1]
```diff
@@ -91,7 +91,7 @@ class BarChart extends React.Component {
               transform={transform}
               xScale={xScale}
               yScale={yScale}
-              xDataType={xDataType}
+              xDataType='date'
```
Comment (bmcfeely) [inline src/components/charts/OperationsHistoryBarChart/OperationsHistoryBarChart.jsx]: is xDataType still needed?

File: src/components/charts/OperationsHistoryBarChart/Rect.jsx [hunk 2]
```diff
@@ -63,15 +59,20 @@ const Rect = (props) => {
           ry={borderRadius}
           data-key={xDataKey}
           data-value={d.y}
-          data-tooltip-x={xScale(d.x) + 5}
+          data-tooltip-current-state={d.currentState}
+          data-tooltip-date={d3.timeFormat('%A %B %-d')(d.label)}
```
Comment (bmcfeely) [inline src/components/charts/OperationsHistoryBarChart/Rect.jsx]: Maybe comment in docs that the three output date formats are fixed ordering and don't change order/layout based on user locale

File: src/components/charts/OperationsHistoryBarChart/OperationsHistoryBarChartTooltip.jsx [hunk 3]
```diff
@@ -0,0 +1,132 @@
+import React from 'react'
+
+const normalIconUri = require('../../../assets/icon_check_outline.svg')
+const warningIconUri = require('../../../assets/icon_eh_warning.svg')
+const criticalIconUri = require('../../../assets/icon_eh_error.svg')
+
+const ToolTip = (props) => {
+  let visibility = 'hidden'
+  let transform = ''
+  let x = 0
```
Comment (bmcfeely) [inline src/components/charts/OperationsHistoryBarChart/OperationsHistoryBarChartTooltip.jsx]: is there a more descriptive name for this?

File: src/components/charts/OperationsHistoryBarChart/OperationsHistoryBarChartTooltip.jsx [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+
+  // TODO Increase width of tooltip if any of the counts are >999, or they will overflow out of the tooltip.
+  let width = 180
+  let height = 70
+  let transformText = 'translate( 10 , 10)'
+  const transformArrow = `translate(${(width / 2) - 20}, ${height})`
+  const transformLine = `translate(${(width / 2) - 20}, ${height})`
+
+  if (props.tooltip.display === true) {
+    const position = props.tooltip.pos
```
Comment (bmcfeely) [inline src/components/charts/OperationsHistoryBarChart/OperationsHistoryBarChartTooltip.jsx]: opinion: inline

File: src/components/charts/OperationsHistoryBarChart/OperationsHistoryBarChartTooltip.jsx [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+        <tspan is x='80' y='10' text-anchor='middle' font-size='12px' font-weight='700' fill='#657d8e'>{props.tooltip.data.date}</tspan>
+        <tspan is x='30' y='28' text-anchor='end' dy='12px' font-size='23px' font-weight={normalFontWeight} fill='#657d8e'>{normalCount}</tspan>
+        <tspan is x='88' y='28' text-anchor='end' dy='12px' font-size='23px' font-weight={criticalFontWeight} fill='#c0392b'>{criticalCount}</tspan>
+        <tspan is x='148' y='28' text-anchor='end' dy='12px' font-size='23px' font-weight={warningFontWeight} fill='#ff9e2c'>{warningCount}</tspan>
+      </text>
+      <image
+        xmlns='http://www.w3.org/2000/svg'
+        x={41}
+        y={34}
+        width={16}
```
Comment (bmcfeely) [inline src/components/charts/OperationsHistoryBarChart/OperationsHistoryBarChartTooltip.jsx]: Just so I learn: why is this 16x16, and other images are 30x30, but they look the same size?

File: src/components/charts/OperationsHistoryBarChart/OperationsHistoryBarChartTooltip.jsx [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+        y={30}
+        width={30}
+        height={30}
+        xlinkHref={criticalIconUri}
+        key={1}
+        className='svgicon'
+        pointerEvents='none'
+        visibility={visibility}
+      />
+      <image
```
Comment (bmcfeely) [inline src/components/charts/OperationsHistoryBarChart/OperationsHistoryBarChartTooltip.jsx]: maybe extract supplier function?

File: src/components/charts/OperationsHistoryBarChart/OperationsHistoryBarChartTooltip.jsx [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+        rx='0'
+        ry='0'
+        visibility={visibility}
+        fill='#ffffff'
+        opacity='1'
+        stroke-opacity='1'
+        stroke={props.tooltip.color}
+        stroke-width='2px'
+      />
+      <line
```
Comment (bmcfeely) [inline src/components/charts/OperationsHistoryBarChart/OperationsHistoryBarChartTooltip.jsx]: just to learn -- where is this line in the image? is this just how we get the text moved down from the top of the box?

File: src/components/charts/OperationsHistoryBarChart/OperationsHistoryBarChart.jsx (hunk 1 above)
Comment (jhegg) [inline src/components/charts/OperationsHistoryBarChart/OperationsHistoryBarChart.jsx]: It's not, thanks!

**Comment by ggascoigne** [comment]
I really like it, but we should just pull it into MP

**Comment by jhegg** [comment]
Cool, going to close this for now.

**Pull request closed without merging**